ios/RCTMapboxGL/RCTMapboxGLManager.m

```diff
@@ -612,22 +612,76 @@ - (NSArray*)serializePacksArray:(NSArray<MGLOfflinePack*>*)packs
+            NSNumber *screenCoordY = pointDict[@"screenCoordY"];
+            NSSet<NSString *> *styleLayerIdentifiers = options[@"layers"];
+
+            CLLocationCoordinate2D location = [_map convertPoint:[sender locationInView:_map] toCoordinateFromView:_map];
+            CGPoint screenCoord = [sender locationInView:_map];
+
+
             CGPoint point = CGPointMake(screenCoordX.floatValue, screenCoordY.floatValue);
 
-            NSArray * features = [mapView visibleFeaturesAtPoint:point];
+            NSArray *features;
+            if (!styleLayerIdentifiers) {
+                features = [mapView visibleFeaturesAtPoint:point];
+            } else {
+                features = [mapView visibleFeaturesAtPoint:point inStyleLayersWithIdentifiers:styleLayerIdentifiers];
+            }
+
+            NSDictionary *geoJSONTypesByMGLClassName = @{
+                    @"MGLPointFeature": @"Point",
+                    @"MGLPolylineFeature": @"LineString",
+                    @"MGLPolygonFeature": @"Polygon"
+            };
 
             // TODO: also accept Rect to use [mapView visibleFeaturesInRect:]
 
-            // TODO: convert features to JS friendly objects
+            // TODO: switch to using MGLFeature's geoJSONFeature method if implemented - https://github.com/mapbox/mapbox-gl-native/issues/6302
+            NSMutableArray *geoJSONFeatures = [[NSMutableArray alloc] init];
+            for (id <MGLFeature> feature in features) {
+//                NSString *geometryType = geoJSONTypesByMGLClassName[NSStringFromClass([feature class])];
+                NSString *geometryType;
+                NSMutableArray *coordinates = [[NSMutableArray alloc] init];
+                if ([feature isKindOfClass:[MGLPointFeature class]]) {
+                    geometryType = @"Point";
+                    coordinates = (NSMutableArray *) @[@(feature.coordinate.longitude), @(feature.coordinate.latitude)];
+                } else if ([feature isKindOfClass:[MGLPolyline class]]) {
+                    MGLMultiPoint *multiPoint = (MGLMultiPoint *)feature;
+                    for (int index = 0; index < multiPoint.pointCount; index++) {
+                        CLLocationCoordinate2D coord = multiPoint.coordinates[index];
+                        [coordinates addObject:@[@(coord.longitude), @(coord.latitude)]];
+                    }
+                    geometryType = @"LineString";
+                } else if ([feature isKindOfClass:[MGLPolygon class]]) {
+                    MGLMultiPoint *multiPoint = (MGLMultiPoint *)feature;
+                    for (int index = 0; index < multiPoint.pointCount; index++) {
+                        CLLocationCoordinate2D coord = multiPoint.coordinates[index];
+                        [coordinates addObject:@[@(coord.longitude), @(coord.latitude)]];
+                    }
+                    geometryType = @"Polygon";
+                }
+                // TODO: checks for MGLMultiPolyline and MGLMultiPolygon
```

Comment:
Also MGLMultiPoint and MGLShapeCollection. Note that MGLPolyline and MGLPolygon currently inherit from MGLMultiPoint (mapbox/mapbox-gl-native#5201), so only handle MGLMultiPoint after those two classes.